File: ios/Quietly/Models/UserBook.swift
import Foundation

struct UserBook: Codable, Identifiable, Hashable {
    let id: UUID
    let userId: UUID
    let bookId: UUID
    var status: ReadingStatus
    var currentPage: Int?
    var rating: Int?
    var startedAt: Date?
    var completedAt: Date?
    let createdAt: Date
    var updatedAt: Date

    // Joined book data
    var book: Book?

    enum CodingKeys: String, CodingKey {
        case id
        case userId = "user_id"
        case bookId = "book_id"
        case status
        case currentPage = "current_page"
        case rating
        case startedAt = "started_at"
        case completedAt = "completed_at"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case book = "books"
    }

    // Computed properties
    var progress: Double {
        guard let pageCount = book?.pageCount, pageCount > 0,
              let current = currentPage else { return 0 }
        return min(Double(current) / Double(pageCount), 1.0)
    }

    var progressPercentage: Int {
        Int(progress * 100)
    }

    var isReading: Bool {
        status == .reading
    }

    var isCompleted: Bool {
        status == .completed
    }
}

// MARK: - Insert Model
struct UserBookInsert: Codable {
    let userId: UUID
    let bookId: UUID
    let status: ReadingStatus
    let currentPage: Int?

    enum CodingKeys: String, CodingKey {
        case userId = "user_id"
        case bookId = "book_id"
        case status
        case currentPage = "current_page"
    }
}

// MARK: - Update Model
struct UserBookUpdate: Codable {
    var status: ReadingStatus?
    var currentPage: Int?
    var rating: Int?
    var startedAt: Date?
    var completedAt: Date?

    enum CodingKeys: String, CodingKey {
        case status
        case currentPage = "current_page"
        case rating
        case startedAt = "started_at"
        case completedAt = "completed_at"
    }
}

File: android/app/src/main/java/com/quietly/app/data/model/UserBook.kt
package com.quietly.app.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class UserBook(
    val id: String,
    @SerialName("user_id")
    val userId: String,
    @SerialName("book_id")
    val bookId: String,
    val status: ReadingStatus,
    @SerialName("current_page")
    val currentPage: Int = 0,
    val rating: Int? = null,
    @SerialName("started_at")
    val startedAt: String? = null,
    @SerialName("finished_at")
    val finishedAt: String? = null,
    @SerialName("created_at")
    val createdAt: String? = null,
    @SerialName("updated_at")
    val updatedAt: String? = null,
    // Joined data
    val book: Book? = null
)
